### ISSUE TITLE
Add YAML support

### ISSUE THREAD

**Issue by jviotti** [sourcemeta/jsonschema#91]

**Comment by UnleashSpirit** [comment]
Need it too, VSCode extension YAML of redhat does it but we need a cli docker tool for CI :)

**Comment by jviotti** [comment]
I'm on vacations at the moment, but I'll take care of it very soon!

**Comment by robbat2** [comment]
another +1 on yaml; to save the `yq` conversion before running `jsonschema`

**Comment by jviotti** [comment]
I promise I'll get into it soon! I gave it a shot back then, but correctly implementing commands like `fmt` on YAML proved to be a bit of a challenge. I think I'll implement it at least for the commands that it's easier to do as a first step. If you are aware of a YAML parser/stringifier in C/C++ that is dependency-free, please let me know!

**Comment by robbat2** [comment]
> I promise I'll get into it soon! I gave it a shot back then, but correctly implementing commands like `fmt` on YAML proved to be a bit of a challenge. I think I'll implement it at least for the commands that it's easier to do as a first step. 

I agree `fmt` is a challenge, but maybe leave that out as a later part. Focusing on reading / convert to JSON / validate.
There are a lot of options & ways to format YAML that are all "correct" but stylistic choices that make more sense in some documents than others (e.g. a short array being inlined, but a longer array being split over lines).


> If you are aware of a YAML parser/stringifier in C/C++ that is dependency-free, please let me know!
Quick research for you:

| project | license | dependencies | notes |
| --- | --- | --- | --- | 
| https://github.com/jbeder/yaml-cpp | MIT | cmake | not maintained anymore? concerns about const-correctness https://github.com/jbeder/yaml-cpp/issues/1275 |
| https://github.com/pantoniou/libfyaml/ | MIT+GPL2+BSD-2-Clause | xxhash bundled | not sure how alive upstream is |
| https://pyyaml.org/wiki/LibYAML | MIT | cmake || make | "stable" |
| https://github.com/tlsa/libcyaml | ISC | libyaml & make | most active library I can see |

**Comment by jviotti** [comment]
@robbat2 Thanks for the help again! I implemented basic YAML interoperability support based in `libyaml` here: https://github.com/sourcemeta/jsontoolkit/tree/main/src/yaml.

The next version of this CLI, which I hope to release in a bit, will include YAML support in most commands, with the exception of `fmt` and the `--fix` option in `lint`. Hopefully we can make those work once I add support for converting back into YAML.

**Comment by robbat2** [comment]
@jviotti thank you for this! It makes a huge improvement in our tooling.

**Comment by jviotti** [comment]
Glad to hear that @robbat2 ! What tooling are you working on btw? Would love to learn more and how we can continue improving this CLI to solve any challenge you have!

**Comment by robbat2** [comment]
This replaces most of https://github.com/RoadieHQ/roadie-backstage-plugins/tree/main/utils/roadie-backstage-entity-validator - converting code-based checks into pure JSONSchema
